controller/src/main/java/io/pravega/controller/store/stream/records/TagRecord.java

```diff
@@ -0,0 +1,135 @@
+        public TagRecordBuilder removeStream(String stream) {
+            this.streams.remove(stream);
+            return this;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    public static TagRecord fromBytes(final byte[] data) {
+        return SERIALIZER.deserialize(data);
+    }
+
+    @SneakyThrows(IOException.class)
+    public byte[] toBytes() {
+        byte[] bytes = SERIALIZER.serialize(this).getCopy();
+        assert bytes.length < TableSegment.MAXIMUM_VALUE_LENGTH : "TagRecord is greater than the TableSegment Maximum value length";
+        return SERIALIZER.serialize(this).getCopy();
+    }
+    
+    private static class TagRecordSerializer
+            extends VersionedSerializer.WithBuilder<TagRecord,
+            TagRecord.TagRecordBuilder> {
+        @Override
+        protected byte getWriteVersion() {
+            return 0;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(0).revision(0, this::write00, this::read00);
+        }
+
+        @Override
+        protected void beforeSerialization(TagRecord streamConfigurationRecord) {
+            Preconditions.checkNotNull(streamConfigurationRecord);
+            Preconditions.checkNotNull(streamConfigurationRecord.getTagName());
+        }
+
+        private void read00(RevisionDataInput revisionDataInput, TagRecord.TagRecordBuilder configurationRecordBuilder)
+                throws IOException {
+            configurationRecordBuilder.tagName(revisionDataInput.readUTF());
+            byte[] comp = revisionDataInput.readArray();
+            Set<String> set = decompressArrayOption(comp);
+            configurationRecordBuilder.streams(set);
+        }
+
+        private void write00(TagRecord streamConfigurationRecord, RevisionDataOutput revisionDataOutput)
+                throws IOException {
+            revisionDataOutput.writeUTF(streamConfigurationRecord.getTagName());
+            byte[] comp = compressArrayOption(streamConfigurationRecord.getStreams());
+            revisionDataOutput.writeArray(comp);
+        }
+
+        @Override
+        protected TagRecord.TagRecordBuilder newBuilder() {
+            return TagRecord.builder();
+        }
+    }
+
+    private static byte[] compressArrayOption(final Set<String> set) throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        OutputStream out = new DeflaterOutputStream(baos);
+        for (String s : set) {
+            out.write(s.getBytes(StandardCharsets.UTF_8));
+            out.write(',');
+        }
+        out.flush();
+        out.close();
+        return baos.toByteArray();
+    }
+
+    private static Set<String> decompressArrayOption(final byte[] compressed) throws IOException {
+        @Cleanup
+        InputStream in = new InflaterInputStream(new ByteArrayInputStream(compressed));
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        IOUtils.copy(in, baos);
```

Comment:
This has a number of unnecessary array copies Wrap `in` in a `DataInputStream` and use `readUTF` (the opposite of what I suggested you do above when you generate this). This will save you the trouble of having to call `IOUtils.copy` and `baos.toString` and `streams.split(...)`. 

Actually you don't even need to encode the comma (separator) in this case as `writeUTF` and `readUTF` take care of that for you.

Reply:
ok, fixed it.